docs/proposal-platforms.md

```diff
@@ -0,0 +1,298 @@
+
+## Example platform configurations
+
+```ini
+[platforms]
+    [[desktop\d\d|laptop\d\d]]
+        # hosts = platform name (default)
+        # Note: "desktop01" and "desktop02" are both valid and distinct platforms
+    [[sugar]]
+        hosts = localhost
+        batch system = slurm
+    [[hpc]]
+        hosts = hpcl1, hpcl2
+        retrieve job logs = True
+        batch system = pbs
+    [[hpcl1-bg]]
+        hosts = hpcl1
+        retrieve job logs = True
+        batch system = background
+    [[hpcl2-bg]]
+        hosts = hpcl2
+        retrieve job logs = True
+        batch system = background
+[platform aliases]
+    [[hpc-bg]]
+        platforms = hpcl1-bg, hpcl2-bg
+```
+
+Note:
+
+* The default for `hosts` is the platform name and platform section headings can
+  be regular expressions to match multiple platform names. This allows us to
+  define many platforms in a single section and handles the case where you have
+  a large number of separate servers which are identical in terms of their
+  platform properties other than their host name.
+
+* The current logic for finding matching host settings is complicated and not
+  fully
+  [documented](https://cylc.github.io/doc/built-sphinx/appendices/site-user-config-ref.html#hosts-host)
+  (sections defining individual host names take precedence over pattern matches
+  regardless of where they come in the order). We propose a different logic for
+  finding matching platform settings:
+  * The pattern will be matched against the full name using
+    [re.fullmatch](https://docs.python.org/3/library/re.html#re.fullmatch)
+    (rather then `re.match` which is currently used for matching hosts).
+  * A hierarchy of platform match expressions from general to specific can be
+    used because config items are processed in the reverse order specified in the
+    file (i.e. bottom up). This is opposite to the method currently used for
+    matching hosts but is required because any additional platforms defined in
+    user config files get added to the end of the list. Currently it is
+    impossible to add hosts defined as regular expressions in the user config if
+    there is already a more generic regular expression defined in the global
+    config.
+  * There will be no special treatment of sections which define a single
+    platform. The first matching section found will be used.
+  * Note that the first reference to a section in the global or user config
+    determines its position in the list. You can repeat a section later in the
+    config and alter some settings but this doesn't change its position.
+
+* The names of all new and existing settings are subject to change until we
+  release cylc 8.0. Initially names will be chosen which match current usage
+  wherever possible. We intend to review all the config settings once we have
+  the key functionality in place.
+
+* The default platform is `localhost`. At cylc 7, any settings for `localhost`
```

Comment:
:+1: Now I have a better understanding of this (existing) behaviour I don't like it. It means that there are 3 possible places a default could have originated from. :disappointed:

Reply:
Not sure I understand your point here. It seems natural to me that some items should have 1) a built-in default - such as batch system defaulting to background - that can overridden by 2) site config, which can be overridden by 3) user global config.  Therefore, for this kind of case the "default" value used by a particular suite could have originated from any of 3 places.  Is this what you mean?  But if you don't like that, what's the alternative? - require every suite to explicitly set every item even when defaults - possibly platform or site specific - are sensible?

Reply:
However, if you're only talking about the way we currently use `localhost` settings as a default for (other) job hosts - which I think is the only way this impacts this proposal? - then I think it's OK to force that to be more explicit.